sdk/identity/identity/src/credentials/managedIdentityCredential/arcMsi.ts

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { AccessToken, GetTokenOptions, RequestPrepareOptions } from "@azure/core-http";
+import { MSI } from "./models";
+import { credentialLogger } from "../../util/logging";
+import { IdentityClient } from "../../client/identityClient";
+import { msiGenericGetToken } from "./utils";
+import { azureArcAPIVersion } from "./constants";
+import { AuthenticationError } from "../../client/errors";
+import { readFile } from "fs";
+
+const logger = credentialLogger("ManagedIdentityCredential - ArcMSI");
+
+// Azure Arc MSI doesn't have a special expiresIn parser.
+const expiresInParser = undefined;
+
+function prepareRequestOptions(resource?: string, clientId?: string): RequestPrepareOptions {
+  const queryParameters: any = {
+    resource,
+    "api-version": azureArcAPIVersion
+  };
+
+  if (clientId) {
+    queryParameters.client_id = clientId;
+  }
+
+  return {
+    // Should be similar to: http://localhost:40342/metadata/identity/oauth2/token
+    url: process.env.IDENTITY_ENDPOINT,
+    method: "GET",
+    queryParameters,
+    headers: {
+      Accept: "application/json",
+      Metadata: true
+    }
+  };
+}
+
+// Since "fs"'s readFileSync locks the thread, and to avoid extra dependencies.
+function readFileAsync(path: string, options: { encoding: string }): Promise<string> {
+  return new Promise((resolve, reject) =>
+    readFile(path, options, (err, data) => {
+      if (err) {
+        reject(err);
+      }
+      resolve(data);
+    })
+  );
+}
+
+async function filePathRequest(
+  identityClient: IdentityClient,
+  requestPrepareOptions: RequestPrepareOptions
+): Promise<string | undefined> {
+  const response = await identityClient.sendRequest(
+    identityClient.createWebResource(requestPrepareOptions)
+  );
+
+  if (response.status !== 401) {
+    throw new AuthenticationError(
+      response.status,
+      "To authenticate with Azure Arc MSI, status code 401 is expected on the first request."
```

Comment:
It might be helpful to users to return some info from the response from the service in the error. I'm wondering if it could have some useful information, for instance something like a malformed request or missing parameter.

Reply:
I'll do that! thank you!